graphix/gflow.py

```diff
@@ -458,3 +574,194 @@ def get_adjacency_matrix(graph):
+    graph: nx.Graph,
+    input: set[int],
+    output: set[int],
+    flow: dict[int, set],
+    meas_planes: dict[int, str] = {},
+) -> bool:
+    """Check whether the flow is valid.
+
+    Parameters
+    ----------
+    graph: nx.Graph
+        graph (incl. in and out)
+    flow: dict[int, set]
+        flow function. flow[i] is the set of qubits to be corrected for the measurement of qubit i.
+    meas_planes: dict[int, str](optional)
+        measurement planes for each qubits. meas_planes[i] is the measurement plane for qubit i.
+
+
+    Returns
+    -------
+    valid_flow: bool
+        True if the flow is valid. False otherwise.
+    """
+
+    valid_flow = True
+    non_outputs = set(graph.nodes) - output
+    # if meas_planes is given, check whether all measurement planes are "XY"
+    for node, plane in meas_planes.items():
+        if plane != "XY" or node not in non_outputs:
+            valid_flow = False
+            return valid_flow
+
+    odd_flow = {node: find_odd_neighbor(graph, corrections) for node, corrections in flow.items()}
+
+    try:
+        layers, depth = get_layers_from_flow(flow, odd_flow, input, output)
+    except ValueError:
+        valid_flow = False
+        return valid_flow
+    node_order = []
+    for d in range(depth):
+        node_order.extend(list(layers[d]))
+    adjacency_matrix, node_list = get_adjacency_matrix(graph)
+    permute = [node_list.index(i) for i in node_order]
+    adjacency_matrix.permute_col(permute)
+    adjacency_matrix.permute_row(permute)
+
+    flow_matrix = MatGF2(np.zeros((len(node_order), len(node_order)), dtype=int))
+    for node, corrections in flow.items():
+        for correction in corrections:
+            row = node_order.index(node)
+            col = node_order.index(correction)
+            # check whether v < f(v)
+            if row >= col:
+                valid_flow = False
+                return valid_flow
+            # check whether the flow arrow is edge of the original graph
+            if adjacency_matrix.data[row, col] != 1:
+                valid_flow = False
+                return valid_flow
+            flow_matrix.data[row, col] = 1
+
+    Neighbor_f = flow_matrix @ adjacency_matrix
+    # Neighbor_f = MatGF2(Neighbor_f.data)
+    # check whether v < N(f(v))
+    triu = np.triu(Neighbor_f.data)
+    triu = MatGF2(triu)
+    dif = triu - Neighbor_f
+    valid_flow &= np.count_nonzero(dif.data) == 0
+    return valid_flow
+
+
+def check_gflow(
+    graph: nx.Graph,
+    input: set[int],
+    output: set[int],
+    gflow: dict[int, set],
+    meas_planes: dict[int, str],
+) -> bool:
+    """Check whether the gflow is valid.
+
+    Parameters
+    ----------
+    graph: nx.Graph
+        graph (incl. in and out)
+    input: set
+        set of node labels for input
+    output: set
+        set of node labels for output
+    gflow: dict[int, set]
+        gflow function. gflow[i] is the set of qubits to be corrected for the measurement of qubit i.
```

Comment:
more explicit definition of `gflow` here, based on original def might be useful?

Reply:
Do you mean [this one](https://github.com/TeamGraphix/graphix/pull/112#issuecomment-1899427717)? We can also place it inside gflow.gflow. Which do you think would be more useful

Reply:
maybe just add 
```python
.. seealso:: :func:`gflow.gflow`
```
 to the docstring, then it's minimally clear what needs to be supplied to the function.